### PR TITLE
Add Serialisation property tests from ouroboros-network

### LINKED ISSUE #1490: move serialisation roundtrips tests from consensus to here
The serialization [roundtrip tests in consensus](https://github.com/input-output-hk/ouroboros-network/blob/ecc60dd9e22a1706125ef337d680a8777ead2a42/ouroboros-consensus-shelley-test/test/Test/Consensus/Shelley/Serialisation.hs#L75) should be moved (along with the [generators](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Generators.hs)) to this repo (`cardano-ledger-specs`).

We can copy them here first and get the tests running. Then the consensus repo can import whatever they need from here.

### PR DESCRIPTION
**Pull request by uroboros**
Closes #1490

Add the serialisation tests from `ouroboros-network` to this repo.
Thanks @mrBliss 🙏

### CODE REVIEW

**Review by JaredCorduan — APPROVED** [review]
looks good, this will be great to have here!

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs [hunk 1]
```diff
@@ -502,395 +511,340 @@ roundTripIpv6 =
...
+    [ testGroup
+        "Serialization Property Tests"
+        [ testGroup
+            "Serialisation roundtrips"
+            [ QC.testProperty "roundtrip Block" prop_roundtrip_Block,
+              QC.testProperty "roundtrip Header" prop_roundtrip_Header,
+              QC.testProperty "roundtrip HeaderHash" prop_roundtrip_HeaderHash,
+              QC.testProperty "roundtrip Tx" prop_roundtrip_Tx,
+              QC.testProperty "roundtrip TxId" prop_roundtrip_TxId,
+              QC.testProperty "roundtrip ApplyTxErr" prop_roundtrip_ApplyTxErr
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs]: Some of these names only exist in consensus, e.g., `HeaderHash` and `ApplyTxErr`. Best to rename to ledger-specs specific ones?

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs [hunk 2]
```diff
@@ -502,395 +511,340 @@ roundTripIpv6 =
...
+            ]
+        ],
+      testGroup
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs]: What's up with all the changes in this file? I would just expect the addition of the `testGroup "Serialization Property Tests"`. Is that just reindenting/relayout? If so, could we avoid that or do it in a different commit?

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs [hunk 3]
```diff
@@ -502,395 +511,340 @@ roundTripIpv6 =
...
-            "txout"
-            (TxOut a (Coin 2))
-            ( T (TkListLen 2)
-                <> S a
-                <> S (Coin 2)
+    "Shelley Serialization Tests"
+    [ testGroup
+        "Serialization Property Tests"
+        [ testGroup
+            "Serialisation roundtrips"
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs]: Do you plan on adding another `testGroup`? If not, I would remove the redundant `testGroup`.

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+prop_roundtrip_HeaderHash = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_Tx :: Mock.Tx -> Property
+prop_roundtrip_Tx = roundtrip' toCBOR ((. Full) . runAnnotator <$> fromCBOR)
+
+prop_roundtrip_TxId :: Mock.TxId -> Property
+prop_roundtrip_TxId = roundtrip toCBOR fromCBOR
+
+prop_roundtrip_ApplyTxErr :: [STS.PredicateFailure Mock.LEDGERS] -> Property
+prop_roundtrip_ApplyTxErr = roundtrip toCBOR fromCBOR
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs]: Again, some of these names are just copy-pasted from consensus and should be updated.

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs:160]: What about all the other types that have en/decoders? For example, `ChainState` and all its subtypes?

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+instance Arbitrary EpochNo where
+  -- Cannot be negative even though it is an 'Integer'
+  arbitrary = EpochNo <$> choose (1, 100000)
+
+instance Arbitrary Mock.Addr where
+  arbitrary =
+    oneof
+      [ Addr <$> arbitrary <*> arbitrary <*> arbitrary
+      -- TODO generate Byron addresses too
+      -- SL.AddrBootstrap
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs:293]: Don't forget about this TODO that was copied.

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs (hunk 2 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs]: yea, it's just indentation, I think, but it does make it very hard to tell if anything actually changed.

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs (hunk 1 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs]: yes thanks, will do

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs (hunk 2 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialization.hs]: this is an unfortunate indentation diff - will see if I can avoid this

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs (hunk 5 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs:160]: yes, I'm adding PrtclState/Ledger State next

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs (hunk 6 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/SerializationProperties.hs:293]: @JaredCorduan see here for when you're done with adding Byron addrs
